Make `DecodeMRPParametersIfPresent()`  and `ParseSigma1()` methods static + fix for TLV reading corner case

Alami-Amine opened the pull request:
- A follow-up to #36679 based on the comment at the bottom of the PR.
- A fix related to #36959
### Changes in the PR:
1. decoupling `DecodeMRPParametersIfPresent()`  from class state and making it static; this is needed to make the factored out TLV Parsing Functions in CASESession and PASESession self-contained and indepedent of class state.

2. Fix for Corner Case in `DecodeMRPParametersIfPresent`:  If the received TLV-encoded SessionParameters structure contains extra TLV elements not known by the receiver (newly added to the specification) but is not terminated with an EndOfContainer element, we must ENSURE that `ExitContainer` is called to ensure that an error is triggered.

3. Added Unit Tests, which also covers above Corner case.

3. Change the name of  `DecodeMRPParametersIfPresent()` to `DecodeSessionParametersIfPresent()` since it is not ONLY decoding MRP but also other Session Parameters (I believe in Matter 1.2 Session Parameters only Included MRP hence the name?)

4. Make `ParseSigma1()` static and completely self-contained.

### Testing

- Added a Unit Test 
-----

              So this is pre-existing, but it looks like the Sigma1 message is partially parsed into parsedSigma1 and partially into the `mRemoteSessionParams` member?

             Followup (do not do this in this PR): we should be passing a `SessionParameters &` to ParseSigma1, and I suspect that method should eventually become a static method that does not touch any class state at all.  There are other places where it does touch state right now, though...

_Originally posted by @bzbarsky-apple in https://github.com/project-chip/connectedhomeip/pull/36679#discussion_r1889527736_

Alami-Amine committed:
Making ParseSigma1 static and not dependent on class state

Alami-Amine committed:
Making DecodeMRPParametersIfPresent self-contained and static and renaming it

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -2324,8 +2326,9 @@ CHIP_ERROR CASESession::ParseSigma1(TLV::ContiguousBufferTLVReader & tlvReader,
-        outParsedSigma1.initiatorMrpParamsPresent = true;
+        ReturnErrorOnFailure(DecodeSessionParametersIfPresent(AsTlvContextTag(Sigma1Tags::kInitiatorSessionParams), tlvReader,
+                                                              outParsedSigma1.initiatorSessionParams));
+        outParsedSigma1.initiatorSessionParamsPresent = true;
```
Should this also set `outParsedSigma1.initiatorSessionParamsPresent` to false if the params are not in fact present?  Is it allowed to reuse a `ParsedSigma1` for multiple calls to this function?  Seems like the answers should be "yes" and "maybe" respectively....

Alami-Amine replied:
I believe that a certain instance of `ParsedSigma1` will never be reused for another call to `ParseSigma1()`, since `ParsedSigma1` is declared on the stack inside `HandleSigma1()` which only calls `ParseSigma1()` once, if we were to call `HandleSigma1()` again, an instance of  `ParsedSigma1` will have gone out of scope, or am I missing something?

bzbarsky-apple replied:
> since ParsedSigma1 is declared on the stack inside HandleSigma1

But the whole point of a separate method like this is that it might be called from places that are _not_ HandleSigma1, no?  But if the intent is that callers must always use a fresh ParsedSigma1 when calling this function, that seems fine but should be documented.

Alami-Amine replied:
@bzbarsky-apple well the point of this method was mostly to improve testability. So all other callers will be Test Code, I assume we shouldn't make such a decision just for Test Code.

Alami-Amine replied:
@tehampson 
1. a related question to your 1st Corner case: is it an issue if we set  mRemoteSessionParams to empty  parameters though? if it isn't, I agree to change to `initiatorSessionParamStructPresent` since its clearer

2. corner case 2: I believe this is already handled within `DecodeSessionParametersIfPresent()` and improved with this PR: since if the old matter device receives an additional TLV element within the SessionParameters that it doesnt recognize: we would get CHIP_NO_ERROR at exit, and then we would call `ExitContainer`, and within `ParseSigma1()` that same sequence will happen. is that what you meant?

tehampson replied:
Essentially, today with your PR, both corner cases I mention do the same thing were it uses the default SessionParameter values. I don't have any strong opinions on this and am okay with how things are done in your PR. But we can technically detect the first one and return an error if we would like. This can also be done in a follow up PR is we are to head in that direction.

@bzbarsky-apple do you have an opinion on what should happen if the initiator has an empty struct (corner case 1?

Essentially my comment on the two corner cases was more a discussion of what should `initiatorSessionParamPresent`, or whatever we end up calling it, be set to and what are we trying to do with it. If in both the corner cases I am talking `initiatorSessionParamPresent` is set to true, then what you have done in the PR is correct. I only brought up the corner case because Boris was asking about when it should be set to false

bzbarsky-apple replied:
The way this stuff works is that when a PairingSession is created its `mRemoteSessionParams` is initialized based on data received from DNS-SD discovery.  See `OperationalSessionSetup::UpdateDeviceData` and `CommissioneeDeviceProxy::UpdateDeviceData`.

Then if the in-band bits have session params, whichever params are present there override whatever was discovered via DNS-SD.

What that means in practice is that empty struct is fine; it just won't override anything, and we will keep using the params we already had.  We could skip the set in that situation too... either way is fine.  Note that PASESession does the set unconditionally, by the way, so maybe we don't need the whole `initiatorSessionParamPresent` thing at all?

Alami-Amine replied:
As discussed privately, it does look like `initiatorSessionParamPresent` flag could be removed, since `mRemoteSessionParams` would anyway be initialised, so if nothing is initiator did not send Session Parameters, we will fallback to the previously set values. I will remove the flag

bzbarsky-apple commented on src/protocols/secure_channel/PairingSession.h:
```diff
@@ -100,6 +100,7 @@ class DLL_EXPORT PairingSession : public SessionDelegate
 
     const ReliableMessageProtocolConfig & GetRemoteMRPConfig() const { return mRemoteSessionParams.GetMRPConfig(); }
     const SessionParameters & GetRemoteSessionParameters() const { return mRemoteSessionParams; }
+    void SetRemoteSessionParameters(const SessionParameters & sessionParams) { mRemoteSessionParams = sessionParams; }
```
Should this be public API?  I wouldn't think so.  Make it protected please.

Alami-Amine replied:
done

Alami-Amine committed:
Making SetRemoteSessionParameters protected

Alami-Amine committed:
Fix for the following Corner case: if the received TLV-encoded SessionParameters struct has extra TLV elements that were added to the spec, Yet the TLV struct is not
terminated with an EndOfContainer Element.
In this Case, we should make sure that ExitContainer so that an Error is triggered.

Alami-Amine committed:
Adding Unit Test for DecodeSessionParametersIfPresent()

tehampson commented on src/protocols/secure_channel/tests/TestPairingSession.cpp:
```diff
@@ -99,22 +97,282 @@ TEST_F(TestPairingSession, PairingSessionTryDecodeMissingMRPParams)
+    if (testCase.testFutureProofing)
+    {
+        uint32_t futureProofTlvElement = 0x777666;
+        writer.Put(ContextTag(8), futureProofTlvElement);
```
ContextTag(8) already exists I thought? `SUPPORTED_TRANSPORTS`. Did you intended to swap these values?

Alami-Amine replied:
`SUPPORTED_TRANSPORTS` is in the Spec but not yet added to the Code, so for now, `ContextTag(8) ` is the future. I thought of adding it as well as `MAX_TCP_MESSAGE_SIZE` so that `futureProofTlvElement` could be `ContextTag(10)`, maybe I should add it to the SessionParameters::Tag enum?

tehampson replied:
Interesting... that is in SessionParameter spec, but it is not included here....

I guess my suggestions is to just not use 8 here but a much higher number so that is something like parsing  SUPPORTED_TRANSPORTS was added to src/messaging/SessionParameters.h that test is a little more robust

Alami-Amine replied:
or maybe I should make it dynamic somehow

Alami-Amine replied:
Yes makes sense, I made the tag=17

Alami-Amine committed:
Fix CI failure related to using aggregate initializer with C-String

Alami-Amine committed:
Integrating Comments
